soroban-env-host/src/native_contract/token/asset_info.rs

```diff
@@ -1,22 +1,27 @@
 use crate::native_contract::token::public_types::AssetInfo;
 use crate::native_contract::token::storage_types::DataKey;
 use crate::{host::Host, HostError};
-use soroban_env_common::{Env, TryIntoVal};
+use soroban_env_common::{Env, StorageType, TryIntoVal};
 
 pub fn write_asset_info(e: &Host, asset_info: AssetInfo) -> Result<(), HostError> {
     let key = DataKey::AssetInfo;
-    e.put_contract_data(key.try_into_val(e)?, asset_info.try_into_val(e)?)?;
+    e.put_contract_data(
```

Comment:
It's not safe because it's what prevents `init_asset` from being called more than once. We could refactor this to check the admin instead and make that Exclusive.